postgrest/postgrest-postgis-spatial-api.md

```diff
@@ -0,0 +1,403 @@
+
+```sh
+db-uri = "postgres://authenticator:gisops@localhost:5432/postgres"
+db-schema = "api"
+db-anon-role = "web_anon"
+server-port = 3000
+```
+
+Great again, now we are ready to start PostgREST.
+
+```sh
+postgrest tutorial.conf
+# or ./postgrest tutorial.conf
+```
+
+You should be able to see something like this:
+
+```sh
+Listening on port 3000
+Attempting to connect to the database...
+Connection successful
+```
+
+The PostgREST server is now ready to serve web requests. 
+But, guess what, there is nothing to serve yet.
+So let's move on and implement some handy spatial API endpoints!
+
+## Step 2 - Let's return a simple GeoJSON Object
+
+As mentioned earlier, we will implement stored procedures which will calculate spatial information on the fly.
+You as a user will have the ability to post `GeoJSON` objects as the payload to the API.
+For the sake of demonstration, we will start with a simple procedure which will return the payload without changing it.
+
+
+Let's bring back our `psql` prompt:
+
+```sh
+sudo docker exec -it tutorial psql -U postgres
+```
+
+We will add a simple function to the `api` schema which can look something like this.
+It takes a single `JSON` argument and returns it again as `JSON`.
+
+
+```sh
+CREATE OR REPLACE FUNCTION api.singlegeojsonparam(single_param json) RETURNS json
+    
+    LANGUAGE sql
+    AS $_$
+    SELECT single_param;
+
+$_$;
+```
+
+Hit enter.
+That's it, let's try it out!
+
+```sh
+curl -X POST \
+  http://localhost:3000/rpc/singlejsonparam \
+  -H 'Content-Type: application/json' \
+  -H 'Prefer: params=single-object' \
+  -d '{"type":"FeatureCollection","features":[{"type":"Feature","geometry":{"type":"Polygon","coordinates":[[[100.0,0.0],[101.0,0.0],[101.0,1.0],[100.0,1.0],[100.0,0.0]]]},"properties":{"prop0":"value1","prop1":{"this":"that"}}},{"type":"Feature","geometry":{"type":"Polygon","coordinates":[[[100.0,0.0],[101.0,0.0],[101.0,1.0],[100.0,1.0],[100.0,0.0]]]},"properties":{"prop0":"value2","prop1":{"this":"that"}}}]}'
+```
+
+It's important to mention here that you can call a function that takes a single parameter of type json by sending the header `Prefer: params=single-object` with your request. 
+That way the JSON request body will be used as the single argument.
+
+
+## Step 3 - Calculating the Length of a LineString
+
+Fantastic, now we can start with some more fun stuff.
+Let's go ahead and write a function which is able to calculate the length of a linestring provided in [EPSG:4326](https://spatialreference.org/ref/epsg/wgs-84/).
+Note, that the GeoJSON we are providing is in degrees, so we will have to transform to pseudo-mercator (or something similar).
+Additionally, we will cast it to `numeric` and after dividing it by 1.000 (to obtain a result in `kilometers`) we will round it to 2 decimals.
+
+
+```sh
+CREATE OR REPLACE FUNCTION api.calc_length(linestring json) RETURNS numeric AS $$ 
+
+    SELECT ROUND(
+      CAST(
```

Comment:
why the CAST to numeric? The result of `ST_Length` is already numeric right..

Reply:
https://stackoverflow.com/questions/13113096/how-to-round-an-average-to-2-decimal-places-in-postgresql/20934099